## Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [14]:
import pandas as pd

In [17]:
columns_names = ['surgery', 'age', 'hospital number', 'rectal temperature', 'pulse',
                'respiratory rate', 'temperature of extremities', 'peripheral pulse', 'mucous membranes',
               'capillary refill time', 'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube',
               'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination - feces',
               'abdomen', 'packed cell volume', 'total protein', 'abdominocentesis appearance',
               'abdomcentesis total protein', 'outcome', 'surgical lesion', 'lesion: site - type - subtype - specific_code',
                 'type of lesion', 'subtype of lesion', 'cp_data']

In [18]:
# Вывод на экран всех столбцов и строк датафрейма
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
table = pd.read_csv('horse_data.csv', names = columns_names, na_values = '?')
table

surgery  age  hospital number  rectal temperature  pulse  \
0        2.0    1           530101                38.5   66.0   
1        1.0    1           534817                39.2   88.0   
2        2.0    1           530334                38.3   40.0   
3        1.0    9          5290409                39.1  164.0   
4        2.0    1           530255                37.3  104.0   
5        2.0    1           528355                 NaN    NaN   
6        1.0    1           526802                37.9   48.0   
7        1.0    1           529607                 NaN   60.0   
8        2.0    1           530051                 NaN   80.0   
9        2.0    9          5299629                38.3   90.0   
10       1.0    1           528548                38.1   66.0   
11       2.0    1           527927                39.1   72.0   
12       1.0    1           528031                37.2   42.0   
13       2.0    9          5291329                38.0   92.0   
14       1.0    1           534917                38.2   76.0   
15       1.0    1           530233                37.6   96.0   
16       1.0    9          5301219                 NaN  128.0   
17       2.0    1           526639                37.5   48.0   
18       1.0    1          5290481                37.6   64.0   
19       2.0    1           532110                39.4  110.0   
20       1.0    1           530157                39.9   72.0   
21       2.0    1           529340                38.4   48.0   
22       1.0    1           521681                38.6   42.0   
23       1.0    9           534998                38.3  130.0   
24       1.0    1           533692                38.1   60.0   
25       2.0    1           529518                37.8   60.0   
26       1.0    1           530526                38.3   72.0   
27       1.0    1           528653                37.8   48.0   
28       1.0    1          5279442                 NaN    NaN   
29       2.0    1           535415                37.7   48.0   
30       2.0    1           529475                37.7   96.0   
31       2.0    1           530242                37.2  108.0   
32       1.0    1           529427                37.2   60.0   
33       1.0    1           529663                38.2   64.0   
34       1.0    1           529796                 NaN  100.0   
35       2.0    1           528812                 NaN  104.0   
36       2.0    1           529493                38.3  112.0   
37       1.0    1           533847                37.8   72.0   
38       2.0    1           528996                38.6   52.0   
39       1.0    9          5277409                39.2  146.0   
40       1.0    1           529498                 NaN   88.0   
41       2.0    9          5288249                39.0  150.0   
42       2.0    1           530301                38.0   60.0   
43       1.0    1           534069                 NaN  120.0   
44       1.0    1           535407                35.4  140.0   
45       2.0    1           529827                 NaN  120.0   
46       1.0    1           529888                37.9   60.0   
47       2.0    1           529821                37.5   48.0   
48       1.0    1           528890                38.9   80.0   
49       2.0    1           529642                37.2   84.0   
50       2.0    1           529766                38.6   46.0   
51       1.0    1           527706                37.4   84.0   
52       2.0    1           529483                 NaN    NaN   
53       2.0    1           530544                38.6   40.0   
54       2.0    1           529461                40.3  114.0   
55       1.0    9          5282839                38.6  160.0   
56       1.0    1           528872                 NaN    NaN   
57       1.0    1           529640                 NaN   64.0   
58       1.0    1           528298                 NaN    NaN   
59       2.0    1           528904                 NaN   96.0   
60       2.0    1           530438      

In [13]:
table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   surgery                                        299 non-null    float64
 1   age                                            300 non-null    int64  
 2   hospital number                                300 non-null    int64  
 3   rectal temperature                             240 non-null    float64
 4   pulse                                          276 non-null    float64
 5   respiratory rate                               242 non-null    float64
 6   temperature of extremities                     244 non-null    float64
 7   peripheral pulse                               231 non-null    float64
 8   mucous membranes                               253 non-null    float64
 9   capillary refill time                          268 non

In [50]:
# Определение основных метрик:
qualitative_columns = ['surgery', 'age', 'peripheral pulse', 'outcome']  # Качественные показатели
quantitative_columns = ['pulse','respiratory rate', 'packed cell volume', 'total protein'] # Количественные показатели

for column in qualitative_columns:
    print(f'Мода столбца {column}: {table[column].mode()[0]}')
print()    
for column in quantitative_columns:
    print(f'Столбец {column}:')
    print(table[column].describe())
    print(f'mode: {table[column].mode()[0]}')
    print()

Мода столбца surgery: 1.0
Мода столбца age: 1
Мода столбца peripheral pulse: 1.0
Мода столбца outcome: 1.0

Столбец pulse:
count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64
mode: 48.0

Столбец respiratory rate:
count    242.000000
mean      30.417355
std       17.642231
min        8.000000
25%       18.500000
50%       24.500000
75%       36.000000
max       96.000000
Name: respiratory rate, dtype: float64
mode: 20.0

Столбец packed cell volume:
count    271.000000
mean      46.295203
std       10.419335
min       23.000000
25%       38.000000
50%       45.000000
75%       52.000000
max       75.000000
Name: packed cell volume, dtype: float64
mode: 37.0

Столбец total protein:
count    267.000000
mean      24.456929
std       27.475009
min        3.300000
25%        6.500000
50%        7.500000
75%       57.000000
max       89.000000
Name: total prot

### Вывод
По данным таблицы чаще всего болели взрослые лошади, перенесшие хирургическое вмешательство. Чаще всего животные выживали. Чаще всего переферический пульс был в норме.

Больныe животные в среднем имели повышенный показатели пульса, минимальное значение соответствует норме, максимальное превышает верхнюю границу в 4,5 раза.
Средний показатель частоты дыхательных движений высокий, максимальное значение превышает норму в 11 раз. 
Размах значений упакованных креток большой: минималное и максимальное значения значительно выходят за границы нормы. Нижний и верхний квартили показывают, что около 50% показателей находтся в пределах нормы.
Размах показателей общего белка так же большой: значения выходят за пределы границ нормы, при этом половина показатлей укладывается в границы нормы.

## Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [52]:
# Определение выбросов в столбцах с количественными данными

for column in quantitative_columns:
    print(column)
    # Расчет межквартильного размаха 
    Q1 = table[column].quantile(0.25)
    Q3 = table[column].quantile(0.75)
    IQR = Q3 - Q1
    
    # Расчет нижнего и врхнего значений выбросов
    lower_outlier = Q1 - (1.5 * IQR)
    higher_outlier = Q3 + (1.5 * IQR)
    print(f'lower_outlier: {lower_outlier}, higher_outlier: {higher_outlier}')
    remove_outliers = table[table[column].between(lower_outlier, higher_outlier, inclusive=True)].sort_values(column)
    outlier = pd.concat([table, remove_outliers]).drop_duplicates(keep=False)
    print(outlier[column].unique())
    print()

pulse
lower_outlier: -12.0, higher_outlier: 148.0
[164.  nan 150. 160. 184.]

respiratory rate
lower_outlier: -7.75, higher_outlier: 62.25
[84. nan 96. 72. 80. 68. 66. 90. 70. 88.]

packed cell volume
lower_outlier: 17.0, higher_outlier: 73.0
[74. nan 75.]

total protein
lower_outlier: -69.25, higher_outlier: 132.75
[nan]



## Вывод:

### Пульс(pulse):
Сердечно-сосудистая система лошади совместима с диапазоном частоты сердечных сокращений от 20 до 240 ударов в минуту. Поэтому имеющие выбросы можно использовать при дальнейшей работе. Возможно, повышенные показатели пульса следует выделить в отдельную группу.

### Дыхательные движения(respiratory rate):
Частота дыхательных движений в норме составляет 8-10, при болезни и повышении нагрузки увеличивается (по некоторым данным под нагрузкой до 40-60 движений в минуту). Кроме того, в сопроводительной файле указано, что полезность информации сомнительна из-за больших колебаний. В это случае значения, превышающие расчетное значение не будут учитываться в дальнейшей работе.


### Объем упакованных клеток(packed cell volume):
Резерв эритроцитов селезенки лошади способен удвоить объем упакованных клеток во время максимальной нагрузки. Если исходить из этого утверждения, выбросы можно использовать в дальнейшемб т.к. значения не намного превышают рассчитанное верхнее значение. 

### Общий белок (total protein):
В столбце нет выбросов, только пропуски значений. 

## Задание 3. Работа с пропусками¶
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [53]:
# Определение количества пропусков

for column in table.columns:
    missings = table[column].isna().mean()
    print(f'{column} - {missings:.1%}')

surgery - 0.3%
age - 0.0%
hospital number - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
respiratory rate - 19.3%
temperature of extremities - 18.7%
peripheral pulse - 23.0%
mucous membranes - 15.7%
capillary refill time - 10.7%
pain - 18.3%
peristalsis - 14.7%
abdominal distension - 18.7%
nasogastric tube - 34.7%
nasogastric reflux - 35.3%
nasogastric reflux PH - 82.3%
rectal examination - feces - 34.0%
abdomen - 39.3%
packed cell volume - 9.7%
total protein - 11.0%
abdominocentesis appearance - 55.0%
abdomcentesis total protein - 66.0%
outcome - 0.3%
surgical lesion - 0.0%
lesion: site - type - subtype - specific_code - 0.0%
type of lesion - 0.0%
subtype of lesion - 0.0%
cp_data - 0.0%


### Выводы:

'surgery' и 'outcome': В каждом столбце прошущено по одному значения, причем в одной и той же строке. В этом случае строка удаляется полностью, т.к. восстановить значения не представляется возможным.

'age': пропусков нет.

'peripheral pulse': пропущено 23% значений. Целесообразно заменить пропуски константой и в дальнейшем анализировать как отдельную категорию.

'pulse':
"концентрация" значений между первым и вторым квартилями больше, чем в остальных. Пропуски заполним модальным занчением, т.к. к указанному диапазону ближе мода.

'respiratory rate' Основная часть показателей находится между нижним и верхним квартилями. Мода и среднее знаечние приблизительно одинаково отстоят от медианы. При этом в диапазоне 28.0 - 40.0 показатели расположены наиболее плотно. К указанному диапазону ближе всего среднее значение, поэтому пропуски будут заполнени именно им.

'packed cell volume': Значения сгруппированы между нижним и верхним квартилями. При этом среднее значение близко к медиане, модальное - значительно отстоит. Поэтому пропуски будут заполнены средним значением.

'total protein': большая часть значений находится до второй квартили. Ко второй квартиле ближе модальное значение, пропуски будут заполнены модой.

In [54]:
# Распределение показателей в столбце 'respiratory rate'
table.groupby('respiratory rate').count()

surgery  age  hospital number  rectal temperature  pulse  \
respiratory rate                                                             
8.0                     1    1                1                   1      1   
9.0                     2    2                2                   2      2   
10.0                    3    3                3                   3      3   
12.0                   19   19               19                  18     18   
13.0                    1    1                1                   0      1   
14.0                    4    4                4                   3      4   
15.0                    1    1                1                   1      1   
16.0                   22   22               22                  18     22   
18.0                    8    8                8                   8      8   
20.0                   27   28               28                  26     27   
21.0                    2    2                2                   2      2   
22.0                    2    2                2                   2      2   
23.0                    1    1                1                   1      0   
24.0                   27   27               27                  22     25   
25.0                    1    1                1                   1      1   
26.0                    1    1                1                   1      1   
28.0                   13   13               13                  12     13   
30.0                   19   19               19                  16     19   
32.0                   11   11               11                   9     11   
34.0                    1    1                1                   1      1   
35.0                    3    3                3                   3      3   
36.0                   16   16               16                  10     16   
40.0                   17   17               17                  15     16   
42.0                    3    3                3                   3      3   
44.0                    3    3                3                   2      3   
48.0                    6    6                6                   5      6   
50.0                    2    2                2                   2      2   
51.0                    2    2                2                   2      2   
52.0                    1    1                1                   1      1   
58.0                    1    1                1                   1      1   
60.0                    4    4                4                   4      4   
66.0                    1    1                1                   1      1   
68.0                    3    3                3                   2      3   
70.0                    2    2                2                   1      2   
72.0                    1    1                1                   1      1   
80.0                    3    3                3                   3      3   
84.0                    2    2                2                   2      2   
88.0                    1    1                1                   1      1   
90.0                    2    2                2                   2      2   
96.0                    2    2                2                   2      2   

                  temperature of extremities  peripheral pulse  \
respiratory rate                                                 
8.0                                        0                 1   
9.0                                        2                 2   
10.0                                       3                 2   
12.0                                      17                16   
13.0                                       1                 1   
14.0                                       2                 2   
15.0                                       1                 0   
16.0                                      19                16   
18.0                                       5                 5   
20.0           

In [57]:
# Удаление строки с пустым значением 'surgery' и 'outcome'
table = table[table['surgery'].notnull()]

# Заполние пропусков в столбцах

table['peripheral pulse'].fillna('constant', inplace = True)
table['pulse'].fillna(table['pulse'].mode()[0], inplace = True)
table['respiratory rate'].fillna(table['respiratory rate'].mean(), inplace = True)
table['packed cell volume'].fillna(table['packed cell volume'].mean(), inplace = True)
table['total protein'].fillna(table['total protein'].mode()[0], inplace = True)

table

surgery  age  hospital number  rectal temperature  pulse  \
0        2.0    1           530101                38.5   66.0   
1        1.0    1           534817                39.2   88.0   
2        2.0    1           530334                38.3   40.0   
3        1.0    9          5290409                39.1  164.0   
4        2.0    1           530255                37.3  104.0   
5        2.0    1           528355                 NaN   48.0   
6        1.0    1           526802                37.9   48.0   
7        1.0    1           529607                 NaN   60.0   
8        2.0    1           530051                 NaN   80.0   
9        2.0    9          5299629                38.3   90.0   
10       1.0    1           528548                38.1   66.0   
11       2.0    1           527927                39.1   72.0   
12       1.0    1           528031                37.2   42.0   
13       2.0    9          5291329                38.0   92.0   
14       1.0    1           534917                38.2   76.0   
15       1.0    1           530233                37.6   96.0   
16       1.0    9          5301219                 NaN  128.0   
17       2.0    1           526639                37.5   48.0   
18       1.0    1          5290481                37.6   64.0   
19       2.0    1           532110                39.4  110.0   
20       1.0    1           530157                39.9   72.0   
21       2.0    1           529340                38.4   48.0   
22       1.0    1           521681                38.6   42.0   
23       1.0    9           534998                38.3  130.0   
24       1.0    1           533692                38.1   60.0   
25       2.0    1           529518                37.8   60.0   
26       1.0    1           530526                38.3   72.0   
27       1.0    1           528653                37.8   48.0   
28       1.0    1          5279442                 NaN   48.0   
29       2.0    1           535415                37.7   48.0   
30       2.0    1           529475                37.7   96.0   
31       2.0    1           530242                37.2  108.0   
32       1.0    1           529427                37.2   60.0   
33       1.0    1           529663                38.2   64.0   
34       1.0    1           529796                 NaN  100.0   
35       2.0    1           528812                 NaN  104.0   
36       2.0    1           529493                38.3  112.0   
37       1.0    1           533847                37.8   72.0   
38       2.0    1           528996                38.6   52.0   
39       1.0    9          5277409                39.2  146.0   
40       1.0    1           529498                 NaN   88.0   
41       2.0    9          5288249                39.0  150.0   
42       2.0    1           530301                38.0   60.0   
43       1.0    1           534069                 NaN  120.0   
44       1.0    1           535407                35.4  140.0   
45       2.0    1           529827                 NaN  120.0   
46       1.0    1           529888                37.9   60.0   
47       2.0    1           529821                37.5   48.0   
48       1.0    1           528890                38.9   80.0   
49       2.0    1           529642                37.2   84.0   
50       2.0    1           529766                38.6   46.0   
51       1.0    1           527706                37.4   84.0   
52       2.0    1           529483                 NaN   48.0   
53       2.0    1           530544                38.6   40.0   
54       2.0    1           529461                40.3  114.0   
55       1.0    9          5282839                38.6  160.0   
56       1.0    1           528872                 NaN   48.0   
57       1.0    1           529640                 NaN   64.0   
58       1.0    1           528298                 NaN   48.0   
59       2.0    1           528904                 NaN   96.0   
60       2.0    1           530438      